File: framework/languages/cpp/e2e/ToActorMessaging/Client/main.cpp
/* SPDX-License-Identifier: FSL-1.1-ALv2 */

#include "../Shared/messages.hpp"

#include <zlink/http_client.hpp>

#include <chrono>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <thread>
#include <vector>

namespace e2e = zlink::e2e::to_actor_messaging;

namespace
{

struct client_configuration_t
{
    std::string actor_http;
    std::string caller_http;
    std::string scenario = "all";
};

client_configuration_t parse_client_configuration (int argc, char **argv)
{
    client_configuration_t configuration;
    for (int index = 1; index < argc; ++index) {
        const std::string argument = argv[index];
        const auto assign = [&] (const std::string &prefix, std::string &target) {
            if (argument.rfind (prefix, 0) != 0) {
                return false;
            }
            target = argument.substr (prefix.size ());
            return true;
        };
        if (!assign ("--actor-http=", configuration.actor_http)
            && !assign ("--caller-http=", configuration.caller_http)
            && !assign ("--scenario=", configuration.scenario)) {
            throw std::runtime_error ("unknown ToActorMessaging client option: " + argument);
        }
    }
    if (configuration.actor_http.empty ()) {
        throw std::runtime_error ("--actor-http is required");
    }
    if (configuration.caller_http.empty ()) {
        throw std::runtime_error ("--caller-http is required");
    }
    if (configuration.scenario.empty ()) {
        throw std::runtime_error ("--scenario must not be empty");
    }
    return configuration;
}

void require (bool condition, const std::string &message)
{
    if (!condition) {
        throw std::runtime_error (message);
    }
}

zlink::http_client::client_t make_http (const std::string &base_url)
{
    return zlink::http_client::client_t::create ()
      .base_url (base_url)
      .timeout (std::chrono::seconds (30))
      .build ();
}

e2e::actor_call_response_t call (zlink::http_client::client_t &caller,
                                 const std::string &endpoint,
                                 const std::string &scenario,
                                 const std::string &actor_id,
                                 const std::string &value)
{
    return caller.post (endpoint)
      .body (e2e::actor_call_request_t{scenario, actor_id, value})
      .fetch<e2e::actor_call_response_t> ();
}

void ensure_actor (zlink::http_client::client_t &actor,
                   const std::string &scenario,
                   const std::string &actor_id)
{
    const auto response = actor.post ("/ensure")
                            .body (e2e::actor_call_request_t{scenario, actor_id, "ensure"})
                            .fetch<e2e::actor_call_response_t> ();
    require (response.error_kind.empty (),
             scenario + " ensure failed: " + response.error_kind);
    require (response.result == "ensured", scenario + " ensure returned " + response.result);
}

void wait_until_ready (zlink::http_client::client_t &caller,
                       const std::string &scenario,
                       const std::string &actor_id)
{
    const auto deadline = std::chrono::steady_clock::now () + std::chrono::seconds (20);
    e2e::actor_call_response_t response;
    while (std::chrono::steady_clock::now () < deadline) {
        response = call (caller, "/request", scenario, actor_id, "ready");
        if (response.error_kind.empty () && response.result == "reply:ready") {
            return;
        }
        if (response.error_kind != "request_failed"
            && response.error_kind != "actor_route_not_found") {
            break;
        }
        std::this_thread::sleep_for (std::chrono::milliseconds (100));
    }
    throw std::runtime_error (scenario + " readiness failed: " + response.error_kind + " "
                              + response.result);
}

void ensure_ready (zlink::http_client::client_t &actor,
                   zlink::http_client::client_t &caller,
                   const std::string &scenario,
                   const std::string &actor_id)
{
    ensure_actor (actor, scenario, actor_id);
    wait_until_ready (caller, scenario + "-ready", actor_id);
}

void assert_call (zlink::http_client::client_t &caller,
                  const std::string &scenario,
                  const std::string &actor_id,
                  const std::string &value,
                  const std::string &expected,
                  bool send)
{
    const auto response = call (caller, send ? "/send" : "/request", scenario, actor_id, value);
    require (response.error_kind.empty (),
             scenario + " unexpected error: " + response.error_kind);
    require (response.result == expected,
             scenario + " expected " + expected + " got " + response.result);
}

void assert_failure (zlink::http_client::client_t &caller,
                     const std::string &scenario,
                     const std::string &actor_id,
                     const std::string &expected_kind,
                     bool send)
{
    const auto response =
      call (caller, send ? "/send" : "/request", scenario, actor_id, "missing");
    require (response.error_kind == expected_kind,
             scenario + " expected " + expected_kind + " got " + response.error_kind
               + " result=" + response.result);
}

void prepare_failure (zlink::http_client::client_t &caller,
                      const std::string &scenario,
                      const std::string &actor_id)
{
    const auto response = call (caller, "/prepare-failure", scenario, actor_id, "prepare");
    require (response.error_kind.empty (),
             scenario + " prepare failed: " + response.error_kind + " " + response.result);
    require (response.result == "prepared", scenario + " prepare returned " + response.result);
}

void require_evidence (const std::vector<e2e::actor_evidence_t> &evidence,
                       const std::string &scenario,
                       const std::string &kind)
{
    for (const auto &entry : evidence) {
        if (entry.scenario == scenario && entry.kind == kind) {
            return;
        }
    }
    throw std::runtime_error (scenario + " " + kind + " evidence missing");
}

std::vector<std::string> split_selector (std::string selector)
{
    if (selector.empty ()) {
        selector = "all";
    }
    std::vector<std::string> scenarios;
    std::stringstream stream (selector);
    std::string item;
    while (std::getline (stream, item, ',')) {
        if (!item.empty ()) {
            scenarios.push_back (item);
        }
    }
    if (scenarios.empty ()) {
        scenarios.push_back ("all");
    }
    return scenarios;
}

bool should_run (const std::vector<std::string> &selected,
                 std::initializer_list<const char *> names)
{
    for (const auto &scenario : selected) {
        if (scenario == "all") {
            return true;
        }
        for (const auto *name : names) {
            if (scenario == name) {
                return true;
            }
        }
    }
    return false;
}

void validate_selector (const std::vector<std::string> &selected)
{
    for (const auto &scenario : selected) {
        if (should_run ({scenario}, {"TA-A1", "ta-a1", "TA-A2", "ta-a2", "TA-A3", "ta-a3",
                                     "TA-A4", "ta-a4", "TA-B1", "ta-b1", "TA-B2", "ta-b2",
                                     "TA-B3", "ta-b3"})) {
            continue;
        }
        throw std::runtime_error ("Unsupported ToActorMessaging scenario: " + scenario);
    }
}

} // namespace

int main (int argc, char **argv)
{
    const auto configuration = parse_client_configuration (argc, argv);
    auto actor = make_http (configuration.actor_http);
    auto caller = make_http (configuration.caller_http);
    const auto selected = split_selector (configuration.scenario);
    validate_selector (selected);

    if (should_run (selected, {"TA-A1", "ta-a1"})) {
        ensure_ready (actor, caller, "TA-A1", "ta-a1");
        assert_call (caller, "TA-A1-send", "ta-a1", "a1-send", "sent", true);
        assert_call (caller, "TA-A1-request", "ta-a1", "a1-request", "reply:a1-request", false);
    }

    if (should_run (selected, {"TA-A2", "ta-a2"})) {
        ensure_ready (actor, caller, "TA-A2", "ta-a2");
        assert_call (caller, "TA-A2-unbound-send", "ta-a2", "a2-send", "sent", true);
        assert_call (caller, "TA-A2-unbound-request", "ta-a2", "a2-request", "reply:a2-request",
                     false);
    }

    if (should_run (selected, {"TA-A3", "ta-a3"})) {
        assert_failure (caller, "TA-A3-before-bind", "ta-a3-missing", "actor_route_not_found",
                        true);
        ensure_ready (actor, caller, "TA-A3", "ta-a3");
        assert_call (caller, "TA-A3-after-bind-send", "ta-a3", "a3-send", "sent", true);
        assert_call (caller, "TA-A3-after-bind-request", "ta-a3", "a3-request",
                     "reply:a3-request", false);
    }

    if (should_run (selected, {"TA-A4", "ta-a4"})) {
        ensure_ready (actor, caller, "TA-A4", "ta-a4");
        assert_call (caller, "TA-A4-disconnected-send", "ta-a4", "a4-send", "sent", true);
        assert_call (caller, "TA-A4-disconnected-request", "ta-a4", "a4-request",
                     "reply:a4-request", false);
        assert_failure (caller, "TA-A4-destroyed", "ta-a4-destroyed", "actor_route_not_found",
                        false);
    }

    if (should_run (selected, {"TA-B1", "ta-b1"})) {
        assert_failure (caller, "TA-B1-missing-send", "missing-actor", "actor_route_not_found",
                        true);
        assert_failure (caller, "TA-B1-missing-request", "missing-actor", "actor_route_not_found",
                        false);
    }

    if (should_run (selected, {"TA-B2", "ta-b2"})) {
        ensure_ready (actor, caller, "TA-B2", "ta-b2-stale");
        prepare_failure (caller, "TA-B2-prepare", "ta-b2-stale");
        assert_failure (caller, "TA-B2-stale-location", "ta-b2-stale", "actor_location_stale",
                        false);
    }

    if (should_run (selected, {"TA-B3", "ta-b3"})) {
        prepare_failure (caller, "TA-B3-prepare", "ta-b3-disconnected");
        assert_failure (caller, "TA-B3-route-not-connected", "ta-b3-disconnected",
                        "route_not_connected", false);
        ensure_ready (actor, caller, "TA-B3", "ta-b3");
        assert_call (caller, "TA-B3-route-restored", "ta-b3", "b3-request", "reply:b3-request",
                     false);
    }

    const auto evidence = actor.get ("/evidence").fetch<std::vector<e2e::actor_evidence_t>> ();
    if (should_run (selected, {"TA-A1", "ta-a1"})) {
        require_evidence (evidence, "TA-A1-send", "send");
        require_evidence (evidence, "TA-A1-request", "request");
    }
    if (should_run (selected, {"TA-A2", "ta-a2"})) {
        require_evidence (evidence, "TA-A2-unbound-send", "send");
    }
    if (should_run (selected, {"TA-A3", "ta-a3"})) {
        require_evidence (evidence, "TA-A3-after-bind-request", "request");
    }
    if (should_run (selected, {"TA-A4", "ta-a4"})) {
        require_evidence (evidence, "TA-A4-disconnected-send", "send");
    }

    std::cout << "to-actor-messaging e2e result=passed\n";
    return 0;
}

File: framework/languages/node/e2e/ToActorMessaging/Client/main.ts
import {
  PacketNames,
  type ActorCallRequest,
  type ActorCallResponse,
  type ActorEvidence,
  type ActorEnsureResponse,
  type ActorPushNotify,
  type ActorRefSnapshot,
  type BindActorReq,
  type BindActorRes
} from '../Shared/messages';
import {
  zlinkStreamConnectorFactory,
  zlinkStreamJsonCodec,
  ZlinkStreamDispatchMode,
  type ZlinkStreamConnector
} from '@zlink-systems/stream-connector';
import { browserE2eArgs, browserE2eFetch, runBrowserE2e } from '../../browser-client-runtime';

interface ClientOptions {
  readonly actorUrl: string;
  readonly callerUrl: string;
  readonly sessionStreamEndpoint: string;
  readonly scenario: string;
}

async function main(): Promise<void> {
  const options = parseClientOptions(browserE2eArgs());
  const scenarios: Record<string, (options: ClientOptions) => Promise<void>> = {
    'TA-A1': runTaA1,
    'TA-A2': runTaA2,
    'TA-A3': runTaA3,
    'TA-A4': runTaA4,
    'TA-B1': runTaB1,
    'TA-B2': runTaB2,
    'TA-B3': runTaB3
  };

  if (options.scenario === 'all') {
    for (const scenario of Object.values(scenarios)) {
      await scenario(options);
    }
  } else {
    const scenario = scenarios[options.scenario];
    if (scenario === undefined) {
      throw new Error(`Unknown scenario '${options.scenario}'.`);
    }
    await scenario(options);
  }

  console.log('to-actor-messaging e2e result=passed');
}

async function runTaA1(options: ClientOptions): Promise<void> {
  const taA1 = await ensureActor(options, 'ta-a1');
  const taA1Session = await bindActor(options, taA1.actor);
  try {
    await assertBoundPush(taA1Session, 'TA-A1-push-before-no-bind', 'ta-a1', 'a1-push-before');
    await assertCall(options, 'TA-A1-send', 'ta-a1', taA1.actor, 'a1-send', 'sent', true);
    await assertCall(options, 'TA-A1-request', 'ta-a1', taA1.actor, 'a1-request', 'reply:a1-request', false);
    await assertBoundPush(taA1Session, 'TA-A1-push-after-no-bind', 'ta-a1', 'a1-push-after');
  } finally {
    await taA1Session.close();
  }
  const evidence = await getJson<ActorEvidence[]>(`${options.actorUrl}/evidence`);
  requireEvidence(evidence, 'TA-A1-send', 'send');
  requireEvidence(evidence, 'TA-A1-request', 'request');
  requireEvidence(evidence, 'TA-A1-push-before-no-bind', 'push');
  requireEvidence(evidence, 'TA-A1-push-after-no-bind', 'push');
  console.log('scenario TA-A1 passed');
}

async function runTaA2(options: ClientOptions): Promise<void> {
  const taA2 = await ensureActor(options, 'ta-a2');
  await assertCall(options, 'TA-A2-unbound-send', 'ta-a2', taA2.actor, 'a2-send', 'sent', true);
  await assertCall(options, 'TA-A2-unbound-request', 'ta-a2', taA2.actor, 'a2-request', 'reply:a2-request', false);
  const evidence = await getJson<ActorEvidence[]>(`${options.actorUrl}/evidence`);
  requireEvidence(evidence, 'TA-A2-unbound-send', 'send');
  requireEvidence(evidence, 'TA-A2-unbound-request', 'request');
  console.log('scenario TA-A2 passed');
}

async function runTaA3(options: ClientOptions): Promise<void> {
  const taA3 = await ensureActor(options, 'ta-a3');
  await assertCall(options, 'TA-A3-before-bind-send', 'ta-a3', taA3.actor, 'a3-before-send', 'sent', true);
  await assertCall(options, 'TA-A3-before-bind-request', 'ta-a3', taA3.actor, 'a3-before-request', 'reply:a3-before-request', false);
  const taA3Session = await bindActor(options, taA3.actor);
  try {
    await assertCall(options, 'TA-A3-after-bind-send', 'ta-a3', taA3.actor, 'a3-send', 'sent', true);
    await assertCall(options, 'TA-A3-after-bind-request', 'ta-a3', taA3.actor, 'a3-request', 'reply:a3-request', false);
    await assertBoundPush(taA3Session, 'TA-A3-after-bind-push', 'ta-a3', 'a3-push');
  } finally {
    await taA3Session.close();
  }
  const evidence = await getJson<ActorEvidence[]>(`${options.actorUrl}/evidence`);
  requireEvidence(evidence, 'TA-A3-before-bind-send', 'send');
  requireEvidence(evidence, 'TA-A3-before-bind-request', 'request');
  requireEvidence(evidence, 'TA-A3-after-bind-send', 'send');
  requireEvidence(evidence, 'TA-A3-after-bind-request', 'request');
  requireEvidence(evidence, 'TA-A3-after-bind-push', 'push');
  console.log('scenario TA-A3 passed');
}

async function runTaA4(options: ClientOptions): Promise<void> {
  const taA4 = await ensureActor(options, 'ta-a4');
  const taA4Session = await bindActor(options, taA4.actor);
  await taA4Session.close();
  await assertCall(options, 'TA-A4-disconnected-send', 'ta-a4', taA4.actor, 'a4-send', 'sent', true);
  await assertCall(options, 'TA-A4-disconnected-request', 'ta-a4', taA4.actor, 'a4-request', 'reply:a4-request', false);
  const evidence = await getJson<ActorEvidence[]>(`${options.actorUrl}/evidence`);
  requireEvidence(evidence, 'TA-A4-disconnected-send', 'send');
  requireEvidence(evidence, 'TA-A4-disconnected-request', 'request');
  console.log('scenario TA-A4 passed');
}

async function runTaB1(options: ClientOptions): Promise<void> {
  const reference = await ensureActor(options, 'ta-b1-reference');
  await postJson(`${options.actorUrl}/actors/ta-b1-reference/destroy`, {});
  const missingActor = reference.actor;
  await assertCall(options, 'TA-B1-missing-send', 'ta-b1-reference', missingActor, 'missing', 'sent', true);
  await assertFailure(
    options,
    'TA-B1-missing-request',
    'ta-b1-reference',
    'actorRouteNotFound',
    false,
    missingActor
  );
  requireNoEvidence(
    await getJson<ActorEvidence[]>(`${options.actorUrl}/evidence`),
    'TA-B1-missing-send'
  );
  console.log('scenario TA-B1 passed');
}

async function runTaB2(options: ClientOptions): Promise<void> {
  const taB2 = await ensureActor(options, 'ta-b2');
  const staleActor = { ...taB2.actor, generation: (BigInt(taB2.actor.generation) + 1n).toString() };
  await assertCall(options, 'TA-B2-stale-send', 'ta-b2', staleActor, 'stale-send', 'sent', true);
  await assertFailure(options, 'TA-B2-stale-request', 'ta-b2', 'actorLocationStale', false, staleActor);
  requireNoEvidence(await getJson<ActorEvidence[]>(`${options.actorUrl}/evidence`), 'TA-B2-stale-send');
  console.log('scenario TA-B2 passed');
}

async function runTaB3(options: ClientOptions): Promise<void> {
  const taB3 = await ensureActor(options, 'ta-b3');
  const disconnectedRouteActor = { ...taB3.actor, nodeRid: 'to-actor-missing-route' };
  await assertCall(options, 'TA-B3-route-not-connected-send', 'ta-b3', disconnectedRouteActor, 'route-send', 'sent', true);
  await assertFailure(options, 'TA-B3-route-not-connected-request', 'ta-b3', 'routeNotConnected', false, disconnectedRouteActor);
  requireNoEvidence(
    await getJson<ActorEvidence[]>(`${options.actorUrl}/evidence`),
    'TA-B3-route-not-connected-send'
  );
  console.log('scenario TA-B3 passed');
}

async function bindActor(options: ClientOptions, actor: ActorRefSnapshot): Promise<ZlinkStreamConnector> {
  const client = zlinkStreamConnectorFactory.create({
    endpoint: options.sessionStreamEndpoint,
    codec: zlinkStreamJsonCodec,
    dispatchMode: ZlinkStreamDispatchMode.Immediate,
    heartbeat: { enabled: false },
    waitTimeoutMs: 10000
  });
  await client.connect();
  const reply = await client
    .request({ actor } satisfies BindActorReq)
    .packetName(PacketNames.bindActor)
    .timeout(5000)
    .submit<BindActorRes>();
  requireCondition(reply.actorId === actor.actorId, `bind ${actor.actorId} actor mismatch.`);
  requireCondition(reply.nodeRid === actor.nodeRid, `bind ${actor.actorId} node mismatch.`);
  requireCondition(reply.generation === actor.generation, `bind ${actor.actorId} generation mismatch.`);
  requireCondition(reply.boundCount >= 1, `bind ${actor.actorId} did not bind any actor.`);
  return client;
}

async function assertBoundPush(
  client: ZlinkStreamConnector,
  scenario: string,
  actorId: string,
  value: string
): Promise<void> {
  const pushed = client.waitFor<ActorPushNotify>(PacketNames.actorPush)
    .where((message) => message.payload.scenario === scenario && message.payload.actorId === actorId)
    .timeout(10000)
    .submit();
  const reply = await client
    .request({ scenario, actorId, value })
    .packetName(PacketNames.actorPush)
    .timeout(5000)
    .submit<{ readonly value: string }>();
  const notify = await pushed;
  requireCondition(reply.value === `pushed:${value}`, `${scenario} reply mismatch.`);
  requireCondition(notify.payload.value === value, `${scenario} push value mismatch.`);
}

async function assertCall(
  options: ClientOptions,
  scenario: string,
  actorId: string,
  actor: ActorRefSnapshot,
  value: string,
  expected: string,
  send: boolean,
  operation?: 'send' | 'request' | 'push'
): Promise<void> {
  const response = await postJson<ActorCallResponse>(
    `${options.callerUrl}/${operation ?? (send ? 'send' : 'request')}`,
    { scenario, actorId, actor, value } satisfies ActorCallRequest
  );
  requireCondition(response.result === expected, `${scenario} expected '${expected}', got '${response.result}'.`);
  requireCondition(response.errorKind === undefined, `${scenario} unexpected error '${response.errorKind}'.`);
}

async function ensureActor(options: ClientOptions, actorId: string): Promise<ActorEnsureResponse> {
  const response = await postJson<ActorEnsureResponse>(`${options.actorUrl}/actors/${actorId}/ensure`, {});
  assertConcreteActorRef(response.actor);
  return response;
}

function assertConcreteActorRef(actor: ActorRefSnapshot): void {
  requireCondition(actor.nodeRid.trim().length > 0, `Actor '${actor.actorId}' node rid is empty.`);
  requireCondition(BigInt(actor.generation) > 0n, `Actor '${actor.actorId}' generation is not positive.`);
}

async function assertFailure(
  options: ClientOptions,
  scenario: string,
  actorId: string,
  expectedKind: string,
  send: boolean,
  actor?: ActorRefSnapshot
): Promise<void> {
  const response = await postJson<ActorCallResponse>(
    `${options.callerUrl}/${send ? 'send' : 'request'}`,
    { scenario, actorId, actor, value: 'missing' } satisfies ActorCallRequest
  );
  requireCondition(response.errorKind === expectedKind, `${scenario} expected '${expectedKind}', got '${response.errorKind}'.`);
}

function requireEvidence(evidence: readonly ActorEvidence[], scenario: string, kind: string): void {
  requireCondition(evidence.some((item) => item.scenario === scenario && item.kind === kind), `${scenario} ${kind} evidence missing.`);
}

function requireNoEvidence(evidence: readonly ActorEvidence[], scenario: string): void {
  requireCondition(
    evidence.every((item) => item.scenario !== scenario),
    `${scenario} unexpectedly reached an actor handler.`
  );
}

async function getJson<T>(url: string): Promise<T> {
  const response = await browserE2eFetch(url);
  if (!response.ok) {
    throw new Error(`${url} failed with ${response.status}`);
  }
  return await response.json() as T;
}

async function postJson<T>(url: string, body: unknown): Promise<T> {
  const response = await browserE2eFetch(url, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(body)
  });
  if (!response.ok) {
    throw new Error(`${url} failed with ${response.status}`);
  }
  return await response.json() as T;
}

function parseClientOptions(args: readonly string[]): ClientOptions {
  const values = new Map<string, string>();
  for (let index = 0; index < args.length; index += 2) {
    const key = args[index]?.replace(/^--/, '');
    const value = args[index + 1];
    if (key === undefined || value === undefined) {
      throw new Error(`Missing value for '${args[index]}'.`);
    }
    values.set(key, value);
  }
  return {
    actorUrl: values.get('actor-url') ?? 'http://127.0.0.1:0',
    callerUrl: values.get('caller-url') ?? 'http://127.0.0.1:0',
    sessionStreamEndpoint: values.get('session-stream-endpoint') ?? 'ws://127.0.0.1:0',
    scenario: values.get('scenario') ?? 'all'
  };
}

function requireCondition(condition: boolean, message: string): void {
  if (!condition) {
    throw new Error(message);
  }
}

void runBrowserE2e('ToActorMessaging', main);
